lib/fastlane/plugin/android_emulator/actions/android_emulator_action.rb

```diff
@@ -16,33 +16,51 @@ def self.run(params)
         system("#{adb} emu kill > /dev/null 2>&1 &")
         sleep(2)
 
-        UI.message("Creating new emulator")
-        FastlaneCore::CommandExecutor.execute(
-          command: "#{sdk_dir}/tools/bin/avdmanager create avd -n '#{params[:name]}' -f -k '#{params[:package]}' -d 'Nexus 5'",
-          print_all: true,
-          print_command: false
-        )
-
-        UI.message("Override configuration")
-        open("#{Dir.home}/.android/avd/#{params[:name]}.avd/config.ini", 'a') { |f|
-          f << "hw.gpu.mode=auto\n"
-          f << "hw.gpu.enabled=yes\n"
-          f << "skin.dynamic=yes\n"
-          f << "skin.name=1080x1920\n"
-        }
-
-        # Verify HAXM installed on mac
-        if FastlaneCore::Helper.mac?
-          kextstat = Actions.sh("kextstat", log: false)
-
-          UI.user_error! "Please install the HAXM-Extension" unless kextstat.include?("intel")
+        avd_available = false
+        config_file = "#{Dir.home}/.android/avd/#{params[:name]}.avd/config.ini"
+        if File.exists?(config_file)
+          avd_image = File.open(config_file).grep(/^image\.sysdir\.1=/)
+          avd_available = avd_image.any? { |s| s.include?(params[:package].gsub(';', '/')) }
+        end
+
+        if params[:retain_previous_avd] && avd_available
+          UI.message("Using existing #{params[:name]} emulator")
+          # wait for additional time for emulator to quit
+          sleep(1)
+        else
+          UI.message("Creating new emulator")
+          FastlaneCore::CommandExecutor.execute(
+            command: "#{sdk_dir}/tools/bin/avdmanager create avd -n '#{params[:name]}' -f -k '#{params[:package]}' -d '#{params[:device_definition]}'",
+            print_all: true,
+            print_command: false
+          )
+
+          UI.message("Override configuration")
+          open("#{Dir.home}/.android/avd/#{params[:name]}.avd/config.ini", 'a') { |f|
+            f << "hw.gpu.mode=auto\n"
+            f << "hw.gpu.enabled=yes\n"
+            f << "skin.dynamic=yes\n"
+            f << "skin.name=1080x1920\n"
+          }
+
+          # Verify HAXM installed on mac
+          if FastlaneCore::Helper.mac?
```

Comment:
In my opinion, that should be outside the "creation block"

Reply:
I agree